File: src/lib/code/mcs-216/session-3/2/2.c
#include <stdio.h>
#include <stdlib.h>

struct Job {
    int id;
    int service_time;
};

// Comparator for qsort: sorts in ascending order of service_time
int compare_jobs(const void *a, const void *b) {
    return ((struct Job*)a)->service_time - ((struct Job*)b)->service_time;
}

int main() {
    struct Job jobs[] = {{1, 5}, {2, 10}, {3, 7}, {4, 8}};
    int n = 4;

    // Sort jobs using Shortest Job First strategy
    qsort(jobs, n, sizeof(struct Job), compare_jobs);

    int finish_time = 0;
    int total_system_time = 0;

    printf("Optimal Sequence: ");
    for (int i = 0; i < n; i++) {
        finish_time += jobs[i].service_time;   // Completion time of job i
        total_system_time += finish_time;     // Cumulative system wait time
        printf("%d%s", jobs[i].id, (i == n - 1) ? "" : " -> ");
    }

    printf("\nTotal Time Spent in System: %d\n", total_system_time);

    return 0;
}

File: src/lib/code/mcs-216/session-3/2/2.rs
#[derive(Debug)]
struct Job {
    id: i32,
    service_time: i32,
}

fn main() {
    let mut jobs = vec![
        Job {
            id: 1,
            service_time: 5,
        },
        Job {
            id: 2,
            service_time: 10,
        },
        Job {
            id: 3,
            service_time: 7,
        },
        Job {
            id: 4,
            service_time: 8,
        },
    ];

    // Sort jobs by service_time ascending
    jobs.sort_by_key(|j| j.service_time);

    let mut finish_time = 0;
    let mut total_system_time = 0;
    let mut sequence = Vec::new();

    for job in jobs {
        finish_time += job.service_time; // Cumulative time for this job
        total_system_time += finish_time; // Total time spent across all jobs
        sequence.push(job.id.to_string());
    }

    println!("Optimal Sequence: {}", sequence.join(" -> "));
    println!("Total Time Spent in System: {}", total_system_time);
}
